maxim/blocks/grid_gating.py

```diff
@@ -47,9 +45,8 @@ def apply(x):
             K.int_shape(x)[3],
         )
         gh, gw = grid_size
-        fh, fw = h // gh, w // gw
 
-        x = BlockImages()(x, patch_size=(fh, fw))
+        x, ph, pw = TFBlockImagesByGrid()(x, grid_size=(gh, gw))
```

Comment:
How come these operations are the same?

Reply:
From the original implementation, the authors implement BlockByGrid by computing the block size of each grid, and using BlockImages (which block images into patches of block-size).

From the paper, the authors explain the difference between "grid" and "block" like that:
![Screenshot from 2022-12-16 12-53-08](https://user-images.githubusercontent.com/8495413/208137664-4433a3b6-c35b-466e-ba71-5afd751b37e0.png)
Note that we can achieve the same result as the grid split by forwarding the block size of [3,2] instead. This is exactly what the authors do in the original code as highlighted [here](https://github.com/sayakpaul/maxim-tf/blob/6364fd25f39b4dcc73565db56d31c2d762376f72/maxim/blocks/misc_gating.py#L119-L121).

They are equivalent because it does the split based on the grid_size as argument instead of the block_size (called as (fh, fw) in the code) as the authors did.

Reply:
Thanks for explaining. 

> Note that we can achieve the same result as the grid split by forwarding a block size of [3,2] instead. 

How is the block size of [3, 2] interpreted in that case?

Reply:
In the original code, it is done as explained, note [here](https://github.com/sayakpaul/maxim-tf/blob/main/maxim/blocks/misc_gating.py#L119-L121):
```
gh, gw = grid_size
fh, fw = h // gh, w // gw
u = BlockImages()(u, patch_size=(fh, fw))
```
Note that this code is very similar to the pseudo-code written [here](https://github.com/sayakpaul/maxim-tf/pull/24#discussion_r1050920798). `grid_size` is passed as a parameter, but `h` and `w`have to be inferred from the image dimensions (which in case of (None, None, 3)), they are None tensors. Thus can't be used in the einops operations.

The way I found to overcome this was to rewrite the operations in tf.

Reply:
Thanks!

So, `TFBlockByGrid()` becomes more idiomatic in that sense. We want to have grid sizes of (2, 2) in the output so, directly pass that as an argument. Correct?

Reply:
Exactly!